File: 650/TwoKeysKeyboard.py
# https://leetcode.com/problems/2-keys-keyboard/description/?envType=daily-question&envId=2024-08-19

class Solution:
    def minSteps(self, n: int) -> int:

        def depthFirstSearch_forNumbers(n, beginString, oldPaste, memoization):
            if len(beginString)>n:
                return 2000000000
            if len(beginString)==n:
                return 0
            if (beginString, oldPaste) in memoization:
                return memoization[(beginString, oldPaste)]
            res=2000000000
            anotherString=beginString+oldPaste
            newPasteString=beginString+beginString
            stringWithOldPaste=1+depthFirstSearch_forNumbers(n, anotherString, oldPaste, memoization)
            stringWithNewPaste=2+depthFirstSearch_forNumbers(n, newPasteString, beginString, memoization)
            res=min(res, min(stringWithOldPaste, stringWithNewPaste))
            memoization[(beginString, oldPaste)]=res
            return memoization[(beginString, oldPaste)]

        if n==1:
            return 0
        beginString="A"
        oldPaste="A"
        memoization={}
        return depthFirstSearch_forNumbers(n, beginString, oldPaste, memoization)+1

File: 6/650/TwoKeysKeyboard.cpp
/*
https://leetcode.com/problems/2-keys-keyboard/description/?envType=daily-question&envId=2024-08-19
*/

class Solution {
public:

    long long dfs(int n, string curr, string oldpst, map<pair<string, string>, int>&memo){
        if(curr.size()>n) return INT_MAX;
        if(curr.size()==n) return 0;
        if(memo.find({curr, oldpst})!=memo.end()) return memo[{curr, oldpst}];
        long long res=INT_MAX;
        string pst=curr+curr;
        long long old=1+dfs(n, curr+oldpst, oldpst, memo);
        long long nwpst=2+dfs(n, pst, curr, memo);
        res=min(res, min(old, nwpst));
        return memo[{curr, oldpst}]=res;
    }

    int minSteps(int n) {
        if(n==1) return 0;
        string curr="A", oldpst="A";
        map<pair<string, string>, int>memo;
        return dfs(n, curr, oldpst, memo)+1;
    }
};